Sprint 20

ArtemDubovitskiy opened the pull request:
Прошу проверить самостоятельную задачу по 20 спринту:

Реализованы Stories с использованием статических данных

Спасибо!)

ArtemDubovitskiy committed:
Добавлено:

- кнопка закрытия просмотра истории
- Ext для размеров ProgressBar
- переименован StoriesView в Preview для отображения на главном экране

ArtemDubovitskiy committed:
Добавлен ProgressBarView

ArtemDubovitskiy committed:
Добавлены StoryView и StoriesView

ArtemDubovitskiy committed:
Изменена модель для историй:

- добавлена модель Stories включающая массив историй;
- исправлены View по новой модели;
- исправлены моковые данные

ArtemDubovitskiy committed:
Добавлен StoriesProgressBarView

ArtemDubovitskiy committed:
Добавлено:

- StoriesTabView
- в StoriesView добавлен StoriesTabView, изменение историй и прогресса

ArtemDubovitskiy committed:
Добавлено:

- обработка выбора истории из превью
- светлый статус бар при просмотре сторис (независимо от темы)

ArtemDubovitskiy committed:
Добавлено:

- Модель Stories дополнена инфо о прочтении
- Поправлено превью историй

ArtemDubovitskiy committed:
Stories перенесены в viewModel

ArtemDubovitskiy committed:
Доработан PreviewStoriesView

ArtemDubovitskiy committed:
Исправлено:

- выбор истории по индексу из массива;
- просмотренные истории отображаются на превью после скрытия историй

So1ovyov commented on TravelSchedule/DesignSystem/Image+Extension.swift:
```diff
@@ -22,7 +22,7 @@ extension Image {
 
     static let changeImage = Image(systemName: "arrow.2.squarepath")
-    static let closeImage = Image(systemName: "multiply.circle.fill")
+    static let closeImage = Image(systemName: "multiply.circle.fill").renderingMode(.original)
```
Рекомендую всегда использовать библиотеки генерации ресурсов типа `SwiftGen`. Избавит от многих проблем при работе с ресурсами(цветами, строками, картинками и прочим)

ArtemDubovitskiy replied:
Спасибо!

So1ovyov commented on TravelSchedule/Extensions/CGFloat+Extension.swift:
```diff
@@ -0,0 +1,13 @@
+
+extension CGFloat {
+    // Stories progress bar
+    static let progressBarCornerRadius: CGFloat = 3
```
Предлагаю не создавать расширение для всего типа `CGFloat`, а использовать отдельный `enum Constants`
```
enum Constants {
   static let progressBarCornerRadius: CGFloat = 3
   static let progressBarHeight: CGFloat = 6
}
```

ArtemDubovitskiy replied:
Спасибо, думал об этом, делал аналогично решению в учебнике.

So1ovyov commented on TravelSchedule/Model/MockData.swift:
```diff
@@ -120,42 +120,97 @@ enum MockData {
+    private static let storyDescription: String = storyTitle + storyTitle + storyTitle
+    
+    static let mockStories: [Stories] = [
+        Stories(
```
Ты уже указал тип данных в массиве, поэтому при объявлении массива достаточно просто вызвать `.init`
```
    static let mockStories: [Stories] = [
        .init(
            stories: [
                Story(image: "Big01",
                      title: storyTitle,
                      description: storyDescription),
                Story(image: "Big02",
                      title: storyTitle,
                      description: storyDescription),
                Story(image: "Big03",
                      title: storyTitle,
                      description: storyDescription),
                Story(image: "Big01",
                      title: storyTitle,
                      description: storyDescription)],
            isRead: false
        ),
        .init(
            stories: [
                Story(image: "Big04",
                      title: storyTitle,
                      description: storyDescription),
                Story(image: "Big05",
                      title: storyTitle,
                      description: storyDescription),
                Story(image: "Big06",
                      title: storyTitle,
                      description: storyDescription)],
            isRead: false
        ),
```

So1ovyov commented on TravelSchedule/Model/Stories.swift:
```diff
@@ -0,0 +1,13 @@
+struct Stories: Hashable, Identifiable {
+    let id = UUID()
+    let stories: [Story]
+    var isRead: Bool
```
Хорошей практикой считается использование иммутабельных структур (все свойства константы, а не переменные). Можно попробовать сделать mutating метод или создать метод, который принимает измененные поля и возвращает новую структуру
```
    let id = UUID()
    let stories: [Story]
    let isRead: Bool
```

ArtemDubovitskiy replied:
Спасибо!

So1ovyov commented on TravelSchedule/View/StartScreen/Stories/Helpers/CloseButton.swift:
```diff
@@ -0,0 +1,29 @@
+
+import SwiftUI
+
+struct CloseButton: View {
```
Хорошее решение декомпозировать код и вынести настройку элемента в отдельную структуру.

So1ovyov commented on TravelSchedule/View/StartScreen/Stories/StoriesView.swift:
```diff
@@ -0,0 +1,84 @@
+                        )
+                    }
+                    
+                    CloseButton(action: {
```
Обрати внимание, можно явно не прописывать параметр `action`
```
CloseButton { 
   storiesIsRead() 
}
```
И это будет работать

ArtemDubovitskiy replied:
Спасибо за ревью!

ArtemDubovitskiy committed:
Исправления по замечаниям ревью:

- в моковых данных тип Stories заменен на .init
- для размеров ProgressBar расширение заменено на enum
- поправлен action для CloseButton
- для StoriesView добавлены состояния (загрузка / контент / ошибка)
